src/migrate/index.js

```diff
@@ -121,22 +131,83 @@ export default class Migrator {
     });
   }
 
+  _createMigrationLockTable(tableName) {
+    return this.knex.schema.createTable(tableName, function(t) {
+      t.integer('is_locked');
+    });
+  }
+
+  _getLockTableName(tableName) {
+    return tableName + '_lock';
+  }
+
+  _isLocked(trx) {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .transacting(trx)
+      .forUpdate()
+      .select('*')
+      .then(data => data[0].is_locked);
+  }
+
+  _lockMigrations(trx) {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .transacting(trx)
+      .update({ is_locked: 1 });
+  }
+
+  _unlockMigrations() {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .update({ is_locked: 0 });
+  }
+
   // Run a batch of current migrations, in sequence.
   _runBatch(migrations, direction) {
-    return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
-      .then(() => this._latestBatchNumber())
-      .then((batchNo) => {
-        if (direction === 'up') batchNo++;
-        return batchNo;
-      })
-      .then((batchNo) => {
-        return this._waterfallBatch(batchNo, migrations, direction)
-      })
-      .catch((error) => {
-        helpers.warn('migrations failed with error: ' + error.message)
-        throw error
-      })
-  }
+    return this.knex.transaction(trx => {
+      return this._isLocked(trx)
+        .then(isLocked => {
+          if (isLocked) {
+            throw new LockError();
+          }
+        })
+        .then(() => this._lockMigrations(trx));
+    })
+    .then(() => Promise.all(_.map(migrations, this._validateMigrationStructure, this)))
+    .then(() => this._latestBatchNumber())
+    .then(batchNo => {
+      if (direction === 'up') batchNo++;
+      return batchNo;
+    })
+    .then(batchNo => {
+      return this._waterfallBatch(batchNo, migrations, direction)
+    })
+    .then(() => this._unlockMigrations())
+    .catch(LockError, error => { 
+      helpers.warn('migrations failed with error: ' + error.message);
+      return error;
+    })
+    .catch(error => {
+      helpers.warn('migrations failed with error: ' + error.message)
+      // If the error was not due to a locking issue, then
+      // remove the lock. Otherwise, leave it in place to
+      // force manual intervention
+      return this._unlockMigrations().finally(function() {
+        return  error;
```

Comment:
if this is `throw error`then one wont need the last then part, which just throws returned error... also `return Promise.reject(error);` would work.